File: python/app/history_store.py
from redis.asyncio import Redis
import json

class HistoryStore:
    """Cache client for managing cache operations"""
    def __init__(self, client: Redis, max_history: int):
        self.client = client
        self.max_history = max_history

    async def get(self, room: str) -> str:
        key = f"history:{room}"
        data = await self.client.lrange(key, 0, -1)
        return [json.loads(message) for message in data]


    async def push(self, room: str, message: dict):
        key = f"history:{room}"
        pipeline = self.client.pipeline()
        pipeline.rpush(key, json.dumps(message))
        pipeline.ltrim(key, -self.max_history, -1)
        await pipeline.execute()
        

File: go/historystore.go
package main

import (
	"github.com/redis/go-redis/v9"
	"context"
	"encoding/json"
)


type Message struct {
	Text string `json:"text"`
	Sender string `json:"sender"`
}


type HistoryStore struct {
	Client *redis.Client
	maxHistory int64
}


func (hs *HistoryStore) get(ctx context.Context, room string) ([]Message, error) {
	key := "history:" + room
	raw, err := hs.Client.LRange(ctx, key, 0, -1).Result()
	if err != nil {
		return nil, err
	}
	messages := make([]Message, 0, len(raw))
	
	for _, stringMessage := range raw {
		var message Message
		rawMessage := []byte(stringMessage)
		err = json.Unmarshal(rawMessage, &message)
		if err != nil {
			return nil, err
		}
		messages = append(messages, message)
	}
	return messages, nil
}


func (hs *HistoryStore) push(ctx context.Context, room string, message Message) error {

	buffer, err := json.Marshal(message)
	if err != nil {
		return err
	}

	key := "history:" + room
	pipeline := hs.Client.Pipeline()
	pipeline.RPush(ctx, key, string(buffer))
	pipeline.LTrim(ctx, key, -hs.maxHistory, -1)
	_, err = pipeline.Exec(ctx)
	return err
}
